generators/beamAndVertexGenerator.cc

```diff
@@ -118,80 +201,201 @@ void beamAndVertexGenerator::randomizeBeamfileStartingPosition() {
-			              beam.particle.mass());
-		}
+		_vertex.SetXYZ(projectedVertexX, projectedVertexY, z);
+		_beam.SetXYZM(_beamMomentumX,
+						_beamMomentumY,
+						_beamMomentumZ,
+						beam.particle.mass());
 	}
 	return true;
 }
 
 
-double beamAndVertexGenerator::getVertexZ(const Target& target) const {
+bool
+beamAndVertexGenerator::eventFromMomentaCorrectMomentumResolution(const Target& target, const Beam& beam) {
+	TRandom3* randomGen = randomNumberGenerator::instance()->getGenerator();
+	if(not loadNextEventFromBeamfile()) return false;
+
+	const double z = getVertexZ(target); // has to be done AFTER loading the next event
+
+	const double dx = _beamMomentumX / _beamMomentumZ;
+	const double dy = _beamMomentumY / _beamMomentumZ;
+	const double projectedVertexX = _vertexX + dx * (z-_vertexZ);
+	const double projectedVertexY = _vertexY + dy * (z-_vertexZ);
+
+	_beamMomentum = sqrt(_beamMomentumX*_beamMomentumX + _beamMomentumY*_beamMomentumY + _beamMomentumZ*_beamMomentumZ);
+	const double truthMomentumPDFSigma = sqrt(1.0/(1.0/_momentumResolution/_momentumResolution+ 1.0/_momentumPDFSigma/_momentumPDFSigma));
+	const double truthMomentumPDFMean = truthMomentumPDFSigma*truthMomentumPDFSigma*( 1./_momentumResolution/_momentumResolution*_beamMomentum
+																					+ 1./_momentumPDFSigma/_momentumPDFSigma*_momentumPDFMean );
+	const double truthMomentum = randomGen->Gaus(truthMomentumPDFMean, truthMomentumPDFSigma);
+
+	_vertex.SetXYZ(projectedVertexX, projectedVertexY, z);
+	const double norm = sqrt(dx*dx + dy*dy + 1);
+	_beam.SetXYZM(truthMomentum * dx / norm,
+	              truthMomentum * dy / norm,
+	              truthMomentum * 1 / norm,
+	              beam.particle.mass());
+	return true;
+}
+
+
+bool
+beamAndVertexGenerator::eventFromInclinations(const Target& target, const Beam& beam) {
 	TRandom3* randomGen = randomNumberGenerator::instance()->getGenerator();
-	double z;
-	do {
-		z = randomGen->Exp(target.interactionLength);
-	} while(z > target.length);
-	z = (target.position.Z() - target.length * 0.5) + z;
-	return z;
+	if(not loadNextEventFromBeamfile()) return false;
+
+	const double z = getVertexZ(target); // has to be done AFTER loading the next event
+
+	TVectorD mean(4);
+	mean[0] = _vertexX;
+	mean[1] = _vertexY;
+	mean[2] = _beamdXdZ;
+	mean[3] = _beamdYdZ;
+	TVectorD sample = multivariantGauss(mean, *_vertexCovarianceMatrix, _sigmaScalingFactor);
+	const double x = sample[0];
+	const double y = sample[1];
+	const double dx = sample[2];
+	const double dy = sample[3];
+	const double norm = sqrt(dx*dx + dy*dy + 1);
+
+
+	const double projectedVertexX = x + dx * (z-_vertexZ);
+	const double projectedVertexY = y + dy * (z-_vertexZ);
+	const double beamMomentum = randomGen->Gaus(_beamMomentum, _sigmaScalingFactor*_beamMomentumSigma);
+		_vertex.SetXYZ(projectedVertexX, projectedVertexY, z);
+		_beam.SetXYZM(  beamMomentum * dx / norm,
+						beamMomentum * dy / norm,
+						beamMomentum * 1  / norm,
```

Comment:
"* 1"? as above

Reply:
Just to make the three lines symmetric.
Should I remove it?